pkg/adsc/delta_adsc.go

```diff
@@ -0,0 +1,1174 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adsc
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	cluster "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	endpoint "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	listener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	route "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/conversion"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/proto"
+	anypb "google.golang.org/protobuf/types/known/anypb"
+	pstruct "google.golang.org/protobuf/types/known/structpb"
+
+	mcp "istio.io/api/mcp/v1alpha1"
+	"istio.io/api/mesh/v1alpha1"
+	mem "istio.io/istio/pilot/pkg/config/memory"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/util"
+	"istio.io/istio/pilot/pkg/serviceregistry/memory"
+	v3 "istio.io/istio/pilot/pkg/xds/v3"
+	"istio.io/istio/pkg/backoff"
+	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/config/constants"
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/config/schema/gvk"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/security"
+	"istio.io/istio/pkg/util/protomarshal"
+	"istio.io/istio/pkg/util/sets"
+	"istio.io/istio/pkg/wellknown"
+)
+
+// DeltaADSC implements a basic client for delta ADS, for use in stress tests and tools
+// or libraries that need to connect to Istio pilot or other ADS servers.
+type DeltaADSC struct {
+	// Stream is the GRPC connection stream, allowing direct GRPC send operations.
+	// Set after Dial is called.
+	stream discovery.AggregatedDiscoveryService_DeltaAggregatedResourcesClient
+
+	// xds client used to create a stream
+	client discovery.AggregatedDiscoveryServiceClient
+	conn   *grpc.ClientConn
+
+	// Indicates if the ADSC client is closed
+	closed bool
+
+	// NodeID is the node identity sent to Pilot.
+	nodeID string
+
+	url string
+
+	watchTime time.Time
+
+	// InitialLoad tracks the time to receive the initial configuration.
+	InitialLoad time.Duration
+
+	// httpListeners contains received listeners with a http_connection_manager filter.
+	httpListeners map[string]*listener.Listener
```

Comment:
I would prefer we make the new client a better client, not just converted from SotW client. The original adsc was designed as a testing client, and is very poorly suited for production use cases. This is what I worked oln in https://github.com/howardjohn/istio/commit/ba3bd91030b5999aca9391a4f395511eacb7cb58

Reply:
Thanks, I've seen https://github.com/istio/istio/issues/36297#issuecomment-1759739439. I'll cherry-pick from what you have and continue with the work.